### ISSUE TITLE
CNI installation leaves temporary files when the istio cni container is killed during copy binaries

### ISSUE THREAD

**Issue by mkralik3** [istio/istio#54311]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

CNI installation leaves temporary files when the istio cni container is killed during copy binaries.
If this situation recurs, the number of leftover temporary files will increase and could eventually fill up the disk space.

More details:
During the installation of CNI, the `copyBinaries` function is [called](https://github.com/istio/istio/blob/master/cni/pkg/install/install.go#L60) which calls [AtomicCopy](https://github.com/istio/istio/blob/master/cni/pkg/install/binaries.go#L43) which calls [AtomicWriteReader](https://github.com/istio/istio/blob/master/pkg/file/file.go#L48) where is a creation of a .tmp file https://github.com/istio/istio/blob/master/pkg/file/file.go#L81. [At the end](https://github.com/istio/istio/blob/master/pkg/file/file.go#L113) of that function, the file is renamed from .tmp to the final name.

When the container is killed or crashes while executing `AtomicWriteReader`, the `defer func()` is not invoked, leaving the temporary file undeleted. If this situation recurs, the number of leftover temporary files will increase and could eventually fill up the disk space.

### Version

```Text
1.20,  but based on the code, it should also happen in the latest version. It is difficult to reproduce because, under normal conditions, the copying of binaries is too fast.
```

**Comment by howardjohn** [comment]
Interesting. Did you find this through a real occurrence, or just from analyzing the code?

**Comment by mkralik3** [comment]
@howardjohn 
Yes, it was a real occurrence  where the cluster node had numerous Istio CNI temporary files (~ 19). However, I haven't been able to reproduce the issue. (Maybe the cluster node's storage was heavily used by something else, causing the Istio-CNI binary copy to take longer, but meanwhile, Kubernetes may have killed the pod. But it is only my "wild" guess.

I discovered this issue during a code analysis (I tried to determine where the .tmp file was created and why it wasn't cleaned up.). With the implemented fix, if the same situation arises, only one temporary file will exist at a time. This will prevent temporary files from consuming all available space, avoiding the risk of cluster instability and a potential snowball effect.
